### PR TITLE
Precompile ECRECOVER

### PR DESCRIPTION
**Pull request by roynalnaruto**
### Description

Add support for verification of the precompile [`ecRecover`](https://www.evm.codes/playground?unit=Wei&codeType=Mnemonic&code='jFirsNplace_parameters%20in%20memoryZ456e9aea5e197a1f1af7a3e85a3212fa4049a3ba34c2289b4c860fc0b0c64ef3whash~Y~28wvX2YZ9242685bf161793cc25603c231bc2f568eb630ea16aa137d2664ac8038825608wrX4YZ4f8ae3bd7535248d0bd448298cc2e2071e56992d0774dc340c368ae950852adawsX6YqqjDo_call~32JSizeX80JOffsetX8VSize~VOffset~1waddressW4QFFFFFFFFwgasqSTATICCALLqqjPut_resulNalonKon_stackqPOPX80qMLOAD'~W1%20w%20jq%5Cnj//%20_%20thKZW32QY0qMSTOREX~0xWqPUSHV0wargsQ%200xNt%20Ke%20Jwret%01JKNQVWXYZ_jqw~_).

### Issue Link

https://github.com/scroll-tech/zkevm-circuits/issues/527

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### Contents

The PR aims at supporting verification of `ecRecover` precompile call.

The PR can be summarised by the following parts:
<details>
<summary><b>Copy Circuit Updates</b></summary>
There are 3 sets of memory operations associated with a precompile call, namely, memory reads from the `caller` to the precompile `call`, memory writes from the precompile call result to the precompile `call` context and lastly, memory writes from the precompile call result to the original `caller` context.

We introduce a new `CopyDataType` for the target precompile, which is the following variant: [`CopyDataType::Precompile(PrecompileCalls)`](https://github.com/scroll-tech/zkevm-circuits/blob/752f4f4af416c5323e7294f83b360067012a5cdb/bus-mapping/src/circuit_input_builder/execution.rs#L198-L199) where [`PrecompileCalls`](https://github.com/scroll-tech/zkevm-circuits/blob/752f4f4af416c5323e7294f83b360067012a5cdb/bus-mapping/src/precompile.rs#L26-L47) is another enum that covers various precompile call types.

The bytes copied to/from precompile target must support the [accumulator check](https://github.com/scroll-tech/zkevm-circuits/blob/752f4f4af416c5323e7294f83b360067012a5cdb/zkevm-circuits/src/copy_circuit.rs#L312-L332).

Since the `CopyDataType` now has 15 variants, we need an additional bit for the `BinaryNumberChip` inside copy circuit. Consequently, this increases the degree of the copy circuit and in order to bring the degree back to `9`, the following changes were had to be made:
- Introduce the following columns https://github.com/scroll-tech/zkevm-circuits/blob/752f4f4af416c5323e7294f83b360067012a5cdb/zkevm-circuits/src/copy_circuit.rs#L65-L73
- Constrain the boolean value assigned to the above columns https://github.com/scroll-tech/zkevm-circuits/blob/752f4f4af416c5323e7294f83b360067012a5cdb/zkevm-circuits/src/copy_circuit.rs#L203-L213
</details>

<details>
<summary><b>Bus Mapping Updates</b></summary>
The memory RWs mentioned above are handled by the [`CallOp`](https://github.com/scroll-tech/zkevm-circuits/blob/752f4f4af416c5323e7294f83b360067012a5cdb/bus-mapping/src/evm/opcodes/callop.rs#L254) opcode.

The `ecRecover` arguments (`[v, r, s, msg_hash, recovered_addr]`) must be pushed to the `CircuitInputBuilder`'s block, so it can later be used as witness data to populate the `SignVerifyTable`. The `SignVerifyTable` is a work-in-progress PR and this PR depends on it.
</details>

<details>
<summary><b>ecRecover Gadget</b></summary>
The difference between signature verification (for tx signature) and elliptic-curve signer recovery is that the signature is not verified in the case of ecRecover. For more details refer "4.1.4 (sig verify)" and "4.1.6 (recovery)" in https://www.secg.org/sec1-v2.pdf.

So in the lookup to the `SigVerifyTable`, `ecRecover` does not care about the `is_valid` column. An example is this [playground setup](https://www.evm.codes/playground?unit=Wei&codeType=Mnemonic&code='jFirsNplace_parameters%20in%20memoryZ456e9aea5e197a1f1af7a3e85a3212fa4049a3ba34c2289b4c860fc0b0c64ef3whash~Y~28wvX2YZ9242685bf161793cc25603c231bc2f568eb630ea16aa137d2664ac8038825608wrX4YZ4f8ae3bd7535248d0bd448298cc2e2071e56992d0774dc340c368ae950852adawsX6YqqjDo_call~32JSizeX80JOffsetX8VSize~VOffset~1waddressW4QFFFFFFFFwgasqSTATICCALLqqjPut_resulNalonKon_stackqPOPX80qMLOAD'~W1%20w%20jq%5Cnj//%20_%20thKZW32QY0qMSTOREX~0xWqPUSHV0wargsQ%200xNt%20Ke%20Jwret%01JKNQVWXYZ_jqw~_) in https://evm.codes. Setting the `calldata_length` to `0x65` (an arbitrary value between 0x60 and 0x80) does recover a signer address without verifying the signature (the signature in this case would have been invalid).

The `ecRecover` gadget is responsible for verifying the internal execution state `PrecompileEcrecover`. The gadget also has `sig_v`, `sig_r`, `sig_s`, `msg_hash`, `recovered_addr` and `recovered` cells to verify the `input` and `output` from the precompile call (this verification is done in the `util::PrecompileGadget`).
</details>

<details>
<summary><b>Precompile Gadget Updates</b></summary>
Add a conditional constraint for `input_rlc` and `output_rlc` (input and output of the precompile call). The `output_rlc` is `0` for the case where no address could be recovered, so the failure case is covered as well. If the address was recovered, then the output RLC is constrained to be equal to the recovered address.
</details>

### How Has This Been Tested?

- [x] Test cases here: https://github.com/scroll-tech/zkevm-circuits/blob/752f4f4af416c5323e7294f83b360067012a5cdb/zkevm-circuits/src/evm_circuit/execution/precompiles/ecrecover.rs#L210
- [x] Testool

### CODE REVIEW

**Comment by lispc** [comment]
we can merge https://github.com/scroll-tech/zkevm-circuits/pull/538 into this branch so SigTable::dev_load can be tested.

After #538 merge, we can merge this.

**Comment by lispc** [comment]
it is time to add the sig lookup? @roynalnaruto

**Comment by roynalnaruto** [comment]
> it is time to add the sig lookup? @roynalnaruto 

Yes, I will add it now.

File: bus-mapping/src/evm/opcodes/callop.rs
Comment (noel2004) [inline bus-mapping/src/evm/opcodes/callop.rs:343]: Some consideration about the `input_len`, which decided the bytes in copydata event:

1. The protocol for precompile call indicate that ' If the input is longer than expected, surplus bytes at the end are ignored.'

2. Every gadget has to prove the input bytes it used is consistent with the bytes in copydata, by applying a rlc of all input bytes and constraint the rlced value equal to `input_bytes_rlc` cell in callop.

3. If the input contain surplus bytes, precompile gadget has to query additional cells to contain these bytes, it maybe a huge cost. For example, the input for modexp gadget can conatin as most as 96 surplus bytes (the up-limit 192 bytes - 96 enforced bytes)

4. Here we have truncated the input length to the up-limit of each precompile contract (if specified). Moreover we may be able to truncate the bytes to just the size it expected (abandon surplus bytes here). Each gadget should be able to prove how many bytes it expected easily.

Comment (roynalnaruto) [reply bus-mapping/src/evm/opcodes/callop.rs:343]: @noel2004 upon checking the `go-ethereum` source code again, the right-padding with 0s is only done for the following precompiles:
- ecrecover
- ecAdd
- ecMul

For `ecPairing` and `blake2f` the input length is already [constrained on the client side](https://github.com/ethereum/go-ethereum/blob/master/core/vm/contracts.go#L536-L538).

At the moment, we are already discarding surplus bytes from calldata.

The pending work is related to right-padding zeroes for the above specified cases, to get:
```rust
// input_rlc is the RLC from all copy-circuit read bytes
let n_pad = expected_input_len - calldata_len;
let final_input_rlc == input_rlc * (pow_of_rand ^ n_pad);
```

**Comment by lispc** [comment]
\b run testool

**Comment by lispc** [comment]
should have a look at the new failures in testool report:

Panic(attempt to subtract with overflow)
Panic(assertion failed: n_padded_zeroes < 128)

**Comment by roynalnaruto** [comment]
\b run testool

**Comment by lispc** [comment]
i noticed the ci takes more time than before (even the "default" mode ci fail to run..).

retrying

File: zkevm-circuits/src/evm_circuit/execution/callop.rs
Comment (lispc) [inline zkevm-circuits/src/evm_circuit/execution/callop.rs]: it is only used for precompiles and should be 0 for non precompiles? so rename it with "precompile_" prefix to avoid mis understanding?

File: zkevm-circuits/src/evm_circuit/execution/precompiles/ecrecover.rs
Comment (lispc) [inline zkevm-circuits/src/evm_circuit/execution/precompiles/ecrecover.rs:83]: `caller_id, call_data_offset, call_data_length, return_data_offset, return_data_length` are not used elsewhere. So here we only read the call_context.   Is it needed really?

File: zkevm-circuits/src/evm_circuit/execution/precompiles/ecrecover.rs
Comment (lispc) [inline zkevm-circuits/src/evm_circuit/execution/precompiles/ecrecover.rs]: make the tests parallel. 16 tests is not a small number
https://github.com/scroll-tech/zkevm-circuits/blob/f24a7956e08bad3ba5e3506f0b007206b67ce7be/zkevm-circuits/src/evm_circuit/execution/callop.rs#L1157

**Review by lispc — APPROVED** [review]
you can merge it when the above 3 comments addressed

File: zkevm-circuits/src/evm_circuit/execution/precompiles/ecrecover.rs
Comment (roynalnaruto) [reply zkevm-circuits/src/evm_circuit/execution/precompiles/ecrecover.rs:83]: Since only a few precompiles are supported at the moment, we are still not aware of which call contexts we would need for which precompile call's exec state validation.

In the bus-mapping crate, we are treating all precompiles in a similar way (at least for the call context lookups) for now. So once we add more precompile support, we can separate the bus-mapping logic into separate modules and then remove call context RWs for individual precompiles wherever they are not required.

**Pull request merged**
